### PR TITLE
[14.0] Shopfloor: Add no_prefill_qty option v2

### PR DESCRIPTION
**Pull request by TDu**
Supersedes #421 

ref: rau-30

### CODE REVIEW

File: shopfloor/components/scan_handler_product.py [hunk 1]
```diff
@@ -15,7 +15,11 @@ class ProductHandler(Component):
     record_type = "product"
 
     def search(self, identifier):
-        return self._search.product_from_scan(identifier)
+        product = self._search.product_from_scan(identifier)
+        if not product:
+            packaging = self._search.packaging_from_scan(identifier)
+            product = packaging.product_id
```
Comment (sebalix) [inline shopfloor/components/scan_handler_product.py:21]: If the goal (not of this PR, but the general idea) is that all scenarios use this scan handler, we do not solve anything by scanning the product+packaging in the same place where some scenarios have a different flow regarding the packaging (like checkout).
We can keep it like this but we need to be aware of that.

File: shopfloor/data/shopfloor_scenario_data.xml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<odoo noupdate="1">
+<odoo noupdate="0">
```
Comment (sebalix) [inline shopfloor/data/shopfloor_scenario_data.xml]: See @simahawk 's comment: https://github.com/OCA/wms/pull/421/files#r1011723561

File: shopfloor/components/scan_handler_product.py (hunk 1 above)
Comment (sebalix) [reply shopfloor/components/scan_handler_product.py:21]: Answer from @simahawk in private: all good like this. Scan handler is used as a global feature on Shopfloor, it should not be used in the code of the specific scenario (for this, use `x_from_scan` methods)

**Comment by TDu** [comment]
The commits related to the improvement of the quantity picker are extracted on https://github.com/OCA/wms/pull/485

File: shopfloor/data/shopfloor_scenario_data.xml (hunk 2 above)
Comment (TDu) [reply shopfloor/data/shopfloor_scenario_data.xml]: Fixed

**Comment by jbaudoux** [comment]
- [x] Zone picking: scan product, then scan invalid pack (with content at an invalid location), error raised "You cannot place it here". The scanned quantity is lost and reset to 0
- [x] Cluster picking: all good
- [x] Checkout: partial pick then scan destination pack => line split is missing (like put in pack is doing)
- [x] Checkout: existing pack, I cannot select a package I created just before that is part of the picking
- [x] Checkout: the new pack button is not enabled (this bug was already there)
- [ ] Checkout: when scanning a second product, only the first one is highlighted (it should also be marked as selected)
- [x] Checkout: edit line quantity, the lines without quantities are not shown anymore after confirmation back on the list of lines

**Comment by TDu** [comment]
The previous commit handles

> Zone picking: scan product, then scan invalid pack (with content at an invalid location), error raised "You cannot place it here". The scanned quantity is lost and reset to 0

And probably many other cases where the quantity being processed was lost

**Comment by TDu** [comment]
Squashed and rebased

File: shopfloor_mobile/static/wms/src/scenario/checkout_states.js [hunk 3]
```diff
@@ -229,14 +229,13 @@ export const checkout_states = function ($instance) {
...
             on_qty_update: (qty) => {
-                console.log(qty);
                 $instance.state.data.qty = qty;
             },
             on_confirm: () => {
                 $instance.wait_call(
                     $instance.odoo.call("set_custom_qty", {
                         picking_id: $instance.state.data.picking.id,
-                        selected_line_ids: $instance.selected_line_ids(),
+                        selected_line_ids: this.selectable_line_ids(),
```
Comment (JuMiSanAr) [inline shopfloor_mobile/static/wms/src/scenario/checkout_states.js]: @TDu that should be `$instance.selectable_line_ids` shouldn't it?

Comment (TDu) [reply shopfloor_mobile/static/wms/src/scenario/checkout_states.js]: Yes, thank you. I added a fixup

**Review by jbaudoux — APPROVED** [review]
I tested and for me it is now good.

**Review by sebalix — APPROVED** [review]

**Comment by sebalix** [comment]
One fixup to squash (maybe other commits as well? Or all good?)

**Review by mmequignon — APPROVED** [review]

**Comment by mmequignon** [comment]
> One fixup to squash (maybe other commits as well? Or all good?)

squashed the fixup commit.

File: shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            move_line,
+            {move_line: origin_qty_done + 1},
+        )
+
+    def test_scan_package_action_scan_product2_to_increment_qty(self):
+        """
+        two lines with the same product
+
+        """
+        # FIXME - test to be fixed
```
Comment (simahawk) [inline shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py]: FIXME or not?

File: shopfloor/tests/test_zone_picking_set_line_destination_no_prefill_qty.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    With the no prefill quantity option set
+
+    * /set_destination
+    """
+
+    def setUp(self):
+        super().setUp()
+        self.service.work.current_picking_type = self.picking2.picking_type_id
+        self.menu.sudo().no_prefill_qty = True
```
Comment (simahawk) [inline shopfloor/tests/test_zone_picking_set_line_destination_no_prefill_qty.py:17]: this should be done on class setup

File: shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        #         "picking_id": picking.id,
+        #         "selected_line_ids": move_lines.ids,
+        #         "barcode": self.product_a.barcode,
+        #     },
+        # )
+        # self._assert_selected_qties(
+        #     response,
+        #     move_lines[0],
+        #     {move_lines[0]: 1, move_lines[1]: 0},
+        # )
```
Comment (sebalix) [inline shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py]: What about this test? Still relevant? What is the issue?

File: shopfloor_mobile/static/wms/src/wms_utils.js [hunk 7]
```diff
@@ -276,13 +278,19 @@ export class WMSUtils {
         return options;
     }
     move_line_qty_picker_options(line, override = {}) {
-        const opts = {
-            init_value: line.quantity,
-            available_packaging: line.product.packaging,
+        // DEPRECATED - To drop in next versions
+        // Use v-bind="utils.wms.move_line_qty_picker_props(...)" instead
+        alert("wms_utils.move_line_qty_picker_options is deprecated.");
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/wms_utils.js]: ALERT? this has to be logged only :thinking:

**Review by simahawk — CHANGES_REQUESTED** [review]
The only blocking point is the on the JS.
That part seems duplicated - partially - from https://github.com/OCA/wms/pull/485

File: shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js [hunk 8]
```diff
@@ -1,7 +1,7 @@
 /**
  * Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
  * @author Simone Orsi <simahawk@gmail.com>
- * Copyright 2021 Jacques-Etienne Baudoux (BCIM)
+ * Copyright 2021-2022 Jacques-Etienne Baudoux (BCIM)
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js]: there's no need to change the (C) year...

File: shopfloor_mobile/static/wms/src/wms_utils.js (hunk 7 above)
Comment (simahawk) [reply shopfloor_mobile/static/wms/src/wms_utils.js]: Please fix this :fire:

**Review by simahawk — CHANGES_REQUESTED** [review]

File: shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js (hunk 8 above)
Comment (jbaudoux) [reply shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js]: was fixed in https://github.com/OCA/wms/pull/485 ...

Comment (jbaudoux) [reply shopfloor_mobile/static/wms/src/components/packaging-qty-picker.js]: fixed here also now

File: shopfloor_mobile/static/wms/src/wms_utils.js (hunk 7 above)
Comment (jbaudoux) [reply shopfloor_mobile/static/wms/src/wms_utils.js]: changed to console.log

File: shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py (hunk 6 above)
Comment (jbaudoux) [reply shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py]: should be re-enabled to see what's failing

File: shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py (hunk 4 above)
Comment (TDu) [reply shopfloor/tests/test_checkout_scan_package_action_no_prefill_qty.py]: Fixed see this commit https://github.com/OCA/wms/pull/483/commits/70190d92a3908a9bfe277360bda6654447c50e9d

**Comment by jbaudoux** [comment]
@JuMiSanAr I spotted some lines in shopfloor_manual_product_transfer_mobile/static/wms/src/scenario/manual_product_transfer.js
that requires utils.wms.move_line_qty_picker_options to be replaced by v-bind=utils.wms.move_line_qty_picker_props
Can you check?

**Comment by sebalix** [comment]
@JuMiSanAr @mmequignon @TDu 2 tests are failing, I can't merge...

**Comment by sebalix** [comment]
All green, let's merge this
/ocabot merge minor

**Pull request merged**